<H2 align="center"> 2019年度知能情報学実験レポート</H2>
<H1 align="center"> 回帰分析$^*$</H1>
<h3 align="center"> F1・F2クラス（水曜日）担当：亀井且有</h3>
<h3 align="center"> F3・F3クラス（木曜日）担当：Ruck THAWONMAS</h3>
<br>
$^*$本件に関する問い合わせは亀井（kamei@ci.ritsumei.ac.jp）まで

<h3 align="left">目的</h3>
<br>
回帰分析の基礎理論を理解し、具体的なデータに対して計算機を用いて回帰分析をする技能を習得する。<br>

<h3 align="left">課題</h3>
<br>
本レポート内の"□"に適当な式、数値、語句を入れ、レポートを完成させた後、manaba+Rで提出せよ。<br>

<h3 align="left">提出締切</h3>
<br>
第1回提出締切：　授業終了時<br>
第2回提出締切：　翌週授業開始前

<h3 align="left">氏名　　：　Luo Yiming</h3>
<h3 align="left">学籍番号：　2600170541-4</h3>
<h3 align="left">クラス　：　F2</h3>
<br>

<h3> 1．　単回帰分析</h3><br>
単回帰モデルは以下の式で表される。<br>
\begin{equation}
y_i =\beta_0+\beta_1x_i+\varepsilon_i \qquad(i=1, \cdots, n; \varepsilon_i \sim \, N(0, \sigma^2))
\end{equation}

回帰式は
\begin{equation}
\hat{y} =\beta_0+\beta_1x
\end{equation}

残差は
\begin{equation}
e_i =y_i-(\beta_0+\beta_1x_i)
\end{equation}

$n$個の残差$e_i$の平方和（残差平方和）　は
\begin{equation}
S_e=\sum_{i=1}^{n}e_i^2
\end{equation}

$S_e$が最小となる切片$\beta_0$と係数$\beta_1$は
\begin{align}
&\beta_0=\bar{y}-\beta_1\bar{x}\\
&\beta_1=\frac{S_{xy}}{S_{xx}}
\end{align}

回帰による平方和は
\begin{equation}
S_R=\frac{S_{xy}^2}{S_{xx}}
\end{equation}

推定した回帰式$\hat{y}$がどの程度有効なのかを表す指標である寄与率は
\begin{equation}
R^2=1-\frac{S_e}{S_{yy}}
\end{equation}

<h3> 2．　重回帰分析</h3>
<h4>2.1 説明変数が2個の場合</h4>
重回帰モデルは以下の式で表される。

\begin{equation}
y_i =\beta_0+\beta_1x_{1i}+\beta_2x_{2i}+\varepsilon_i \qquad(i=1, \cdots, n; \varepsilon_i \sim \, N(0, \sigma^2))
\end{equation}

回帰式は
\begin{equation}
\hat{y} =\beta_0+\beta_1x_{1i}+\beta_2x_{2i}
\end{equation}

残差平方和$S_e$が最小となる切片$\beta_0$および係数$\beta_1,\beta_2$は

\begin{align}
    &\beta_0=\bar{y}-\beta_1\bar{x_1}-\beta_2\bar{x_2}\\
    &\beta_1=\frac{S_{22}S_{1y}-S_{12}S_{2y}}{S_{11}S_{22}-S_{12}^2}\\
    &\beta_2=\frac{-S_{12}S_{1y}+S_{11}S_{2y}}{S_{11}S_{22}-S_{12}^2}
\end{align}

回帰による平方和は
\begin{equation}
S_R=\beta_1S_{1y}+\beta_2S_{2y}
\end{equation}

推定した回帰式$\hat{y}$がどの程度有効なのかを表す指標である寄与率は

\begin{equation}
R^2=1-\frac{S_3}{S_{yy}}
\end{equation}

<h4>2.2 説明変数が3個の場合</h4>
重回帰モデルは以下の式で表される。

\begin{equation}
y_i =\beta_0+\beta_1x_{1i}+\beta_2x_{2i}+\beta_3x_{3i}+\varepsilon_i \qquad(i=1, \cdots, n; \varepsilon_i \sim \, N(0, \sigma^2))
\end{equation}

回帰式は

\begin{equation}
\hat{y}_i =\beta_0+\beta_1x_{1i}+\beta_2x_{2i}+\beta_3x_{3i}
\end{equation}

残差平方和$S_e$が最小となる切片は

\begin{equation}
\beta_0=\bar{y}-\beta_1\bar{x_1}-\beta_2\bar{x_2}-\beta_3\bar{x_3}
\end{equation}

$\beta_1,\,\beta_2,\,\beta_3$は次式で与えられる。

\begin{equation}
\left(
\begin{array}{c}
\beta_1\\
\beta_2\\
\beta_3
\end{array}
\right) =
\left(
\begin{array}{ccc}
S_{11}  & S_{12}   & S_{13}  \\
S_{21}  & S_{22}   & S_{23}  \\
S_{31}  & S_{32}   & S_{33} 
\end{array}
\right)^{-1}
\left(
\begin{array}{c}
S_{1y}\\
S_{2y}\\
S_{3y}\\
\end{array}
\right)
\end{equation}

回帰による平方和は

\begin{equation}
S_R=\beta_1S_{1y}+\beta_2S_{2y}+\beta_3S_{3y}
\end{equation}

寄与率は

\begin{equation}
R^2=1-\frac{S_R}{S_{yy}}=r_{y\hat{y}}^2
\end{equation}

<h3> 3．　エクセルを用いた重回帰分析</h3>

<ol>
<li>manaba+Rより下記のサンプルデータファイルを取得せよ。</li>
<ul>
    <li>F1クラス：data_ra_F1.xlsx</li>
    <li>F2クラス：data_ra_F2.xlsx</li>
    <li>F3クラス：data_ra_F3.xlsx</li>
    <li>F4クラス：data_ra_F4.xlsx</li>
</ul>
<br>

<li>サンプルデータのタブ「重回帰データ」の$x_1, x_2, x_3,y$より$x$の分散共分散行列$\Sigma_x^{(3\times3)}$および$x$と$y$の共分散行列$\Sigma_{xy}^{(3\times1)}$を求めよ$^1$。</li>

\begin{align}
\left(
\begin{array}{ccc}
S_{11}  & S_{12}   & S_{13}  \\
S_{12}  & S_{22}   & S_{23}  \\
S_{13}  & S_{23}   & S_{33}  
\end{array}
\right) &=
\left(
\begin{array}{ccc}
2.0144  & 3.7094   & -0.5706  \\
3.7094  & 47.5044   & -11.7756 \\
-0.5706  & -11.7756   & 4.9444
\end{array}
\right)\\
\left(
\begin{array}{c}
S_{1y}\\
S_{2y}\\
S_{3y}
\end{array}
\right) &=
\left(
\begin{array}{c}
7.7774\\
42.8724\\
-11.0076
\end{array}
\right)
\end{align}

<li>分散共分散行列$\Sigma_x$の逆行列$\Sigma_x^{-1}$を求めよ$^2$。</li>

\begin{align}
\left(
\begin{array}{ccc}
S_{11}  & S_{12}   & S_{13}  \\
S_{12}  & S_{22}   & S_{23}  \\
S_{13}  & S_{23}   & S_{33}  
\end{array}
\right) ^{-1}&=
\left(
\begin{array}{ccc}
0.6007278  & -0.07255986   & -0.10348283  \\
-0.07255986  & 0.061528   & 0.13488646  \\
-0.10348283  & 0.13488646   & 0.51155282 
\end{array}
\right)
\end{align}

<li>$\beta_0,\beta_1,\beta_2,\beta_3$を求めよ。</li>

\begin{equation}
\left(
\begin{array}{c}
\beta_0\\
\beta_1\\
\beta_2\\
\beta_3
\end{array}
\right) =
\left(
\begin{array}{c}
-11.6481961\\
2.70038342\\
0.52979155\\
-0.65288969
\end{array}
\right) 
\end{equation}


<li>回帰による平方和は</li>

\begin{equation}
S_R=50.902146
\end{equation}

<li>寄与率は</li>

\begin{equation}
R^2=0.84274518
\end{equation}

<br><br>
$^1$  分散・共分散は関数VAR.P, COVARIANCE.PあるいはVAR.S, COVARIANCE.Sを使用するが、.Pと.Sの違いに注意すること。<br>
$^2$  行列の積や逆行列の求め方はhttps://bellcurve.jp/statistics/blog/15368.htmlを参照。

<h3> 4．　Pythonを用いた重回帰分析</h3>

<ol>
    <li>manaba+Rより下記のサンプルデータファイルを取得せよ。</li>
    <ul>
        <li>F1クラス：data_ra_F1.csv</li>
        <li>F2クラス：data_ra_F2.csv</li>
        <li>F3クラス：data_ra_F3.csv</li>
        <li>F4クラス：data_ra_F4.csv</li>
   </ul><br>
    <li>Pythonでコーディングをして、サンプルデータを重回帰分析を実行せよ。</li>
</ol>

In [1]:
import pandas as pd
import numpy as np

data=pd.read_csv("data_ra_F2.csv", sep=",")
print(data.head())

from sklearn import linear_model
clf=linear_model.LinearRegression()

data_except_y=data.drop("y",axis=1)
X=data_except_y.values
y=data['y'].values

clf.fit(X, y)

print(pd.DataFrame({"Name":data_except_y.columns,
    "Coefficients":clf.coef_}).sort_values(by='Coefficients'))
print("SEPPEN=",clf.intercept_)
print("KIYORITSU=",clf.score(X, y))

     x1    x2    x3     y
0  15.7  30.2  19.7  38.0
1  15.9  25.9  20.4  31.7
2  17.9  21.4  20.9  35.2
3  14.6  30.6  20.1  27.4
4  16.8  44.8  14.3  51.6
   Coefficients Name
2     -0.652890   x3
1      0.529792   x2
0      2.700383   x1
SEPPEN= -11.648196102536268
KIYORITSU= 0.8427451800497451


<ol>
    <li　value="3">切片および$x_1, x_2, x_3$の係数がエクセルを用いた結果$\beta_0, \beta_1, \beta_2, \beta_3$と一致することを確認せよ。</li>
    <li>寄与率がエクセルを用いた結果$R^2$と一致することを確認せよ。</li>
</ol>

<h3> 5．　Pythonを用いた重回帰分析（発展課題）</h3>
<br>
manaba+Rよりサンプルデータ（data_ra_winequality-red.csv}）を取得し、'quality'を目的変数、それ以外を説明変数、としたときの回帰式における各説明変数の係数および切片を求めよ。

In [2]:
import pandas as pd
import numpy as np

wine=pd.read_csv("data_ra_winequality-red.csv", sep=",")
print(wine.head())

from sklearn import linear_model
clf=linear_model.LinearRegression()

wine_except_quality=wine.drop("quality", axis = 1)
X=wine_except_quality.values
y=wine['quality'].values

clf.fit(X, y)

print(pd.DataFrame({"Name":wine_except_quality.columns,"Coefficients":clf.coef_}).sort_values(by='Coefficients'))
print("SEPPEN=",clf.intercept_)
print("KIYORITSU=",clf.score(X, y))


   fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5 

プログラムが正しく動作すれば、出力は以下となる。
<pre>
fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \ 
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075  
4            7.4              0.70         0.00             1.9      0.076  

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \ 
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56  

   alcohol  quality 
0      9.4        5 
1      9.8        5 
2      9.8        5 
3      9.8        6 
4      9.4        5  
                    Name  Coefficients 
7                density    -17.881164
4              chlorides     -1.874225
1       volatile acidity     -1.083590
8                     pH     -0.413653
2            citric acid     -0.182564
6   total sulfur dioxide     -0.003265
5    free sulfur dioxide      0.004361
3         residual sugar      0.016331
0          fixed acidity      0.024991
10               alcohol      0.276198
9              sulphates      0.916334
SEPPEN= 21.965208449451936
KIYORITSU= 0.3605517030386882
</pre>